Plugins: redirect to WooCommerce onboarding once plugin is installed

retrofox opened the pull request:
#### Changes proposed in this Pull Request

This PR performs a client redirect to the WooCommerce onboarding (wp-admin) once the plugin is installed.

#### Testing instructions

Test with **JETPACK**  site.

* Go to WooCommerce plugin page in calypso: `<hostname>/plugins/woocommerce/<site-id>`

<img width="1078" alt="Screen Shot 2021-04-26 at 9 30 50 AM" src="https://user-images.githubusercontent.com/77539/116083029-8606b080-a672-11eb-987e-0a6edc481c81.png">

* Click on the Install button to trigger the install process.
* Once the plugin is installed, the client should redirect to the WooCoomerce onboarding page (wp-admin)

<img width="1081" alt="Screen Shot 2021-04-26 at 9 34 57 AM" src="https://user-images.githubusercontent.com/77539/116083166-b3ebf500-a672-11eb-9765-3854c621daab.png">



Related to https://github.com/Automattic/wp-calypso/issues/51327

retrofox committed:
selectors: add getSiteWoocommerceWizardUrl()

tyxla reviewed (commented):
Instead of polluting the (already pretty polluted) single plugin component with this logic and further relying on component lifecycle, could we just do the redirect [directly in Redux](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/plugins/installed/actions.js#L390) upon a successful plugin installation?

retrofox commented:
> Instead of polluting the (already pretty polluted) single plugin component with this logic and further relying on component lifecycle, could we just do the redirect [directly in Redux](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/plugins/installed/actions.js#L390) upon a successful plugin installation?

That looks more straightforward and the processing level (data flow) seems to be properer. Let'me re-implement the PR. Thanks for your help.

tyxla commented:
> > Instead of polluting the (already pretty polluted) single plugin component with this logic and further relying on component lifecycle, could we just do the redirect [directly in Redux](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/plugins/installed/actions.js#L390) upon a successful plugin installation?
> 
> That looks more straightforward and the processing level (data flow) seems to be properer. Let'me re-implement the PR. Thanks for your help.

Sure thing, let me know when it's ready for review, I'm happy to take another look. I spent a bunch of time in the plugins area of the code over the past year due to their reduxification, so I feel at home in this part of the code 😉

retrofox committed:
plugins: redirects to woocommerce just installed

retrofox committed:
plugins: redirect to woo wizard page

retrofox commented:
> Sure thing, let me know when it's ready for review, I'm happy to take another look.

I really appreciate it. PR is rebased, removed all stuff at the component level, and all processing is in the data layer. I've added another function called just right after the plugin is successfully installed.

> I spent a bunch of time in the plugins area of the code over the past year due to their reduxification, so I feel at home in this part of the code 😉

🎼 Music to my ears👂

retrofox commented:
> could we just do the redirect [directly in Redux](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/plugins/installed/actions.js#L390) upon a successful plugin installation?

Thinking well... I'm not sure whether this is the proper approach because it depends on the current page. Usually, the user installs the plugin (woocommerce) from the plugin single page:

```
<host>/plugins/woocommerce/<site>
```

However, technically it's possible to trigger the installing process from any part of the code. For instance, we could add an install/uninstall button for every card in the plugins page:

![image](https://user-images.githubusercontent.com/77539/116296804-08759a00-a771-11eb-8f35-cdfe6a1cc3f2.png)

In that case, the app will keep redirecting to the WooCommerce wizard (wp-admin) even from this page, because the redirect handler only checks the plugin slug (woocommerce). We could check the current view, but it starts to feel weird in the data layer.

Thoughts?

tyxla commented:
I understand, that makes sense 👍  Thanks for elaborating @retrofox!

In that case, could we move this functionality to a functional higher-order component and use hooks for it? Then we could wrap all "interested" components with it (`SinglePlugin` in this case), avoiding clutter, also avoiding reliance on `SinglePlugin`'s component lifecycle, while still separating concerns from the convoluted `SinglePlugin` component logic. 

What do you think?

retrofox commented:
> What do you think?

Sounds good to me. :+1:

retrofox commented:
> What do you think?

https://github.com/Automattic/wp-calypso/pull/52398 is the new PR :-)

retrofox commented:
I'm going to close this one in favor of https://github.com/Automattic/wp-calypso/pull/52398
